### ISSUE TITLE
consistency: redeploy EVCs that are enabled and inactive and now have paths to be deployed

### ISSUE THREAD

**Issue by italovalcy** [kytos-ng/mef_eline#395]
Hi,

Suppose that for some reason, an EVC is not deployed when a Link UP that could benefit it is received (I do have some use cases, but I would like to leave the generic discussion here, ie, just consider that the Link UP event was somehow lost). In this case, the mef_eline consistency check is not checking the EVC, due to this condition (`and (circuit.is_intra_switch() or circuit.current_path) `):

```
    @staticmethod
    def should_be_checked(circuit):
        "Verify if the circuit meets the necessary conditions to be checked"
        # pylint: disable=too-many-boolean-expressions
        if (
                circuit.is_enabled()
                and not circuit.is_active()
                and not circuit.lock.locked()
                and not circuit.has_recent_removed_flow()
                and not circuit.is_recent_updated()
                # if a inter-switch EVC does not have current_path, it does not
                # make sense to run sdntrace on it
                and (circuit.is_intra_switch() or circuit.current_path)
                ):
            return True
        return False
```

When this happens, together with situations like described in #394 , the EVC will stay inactive until a new event related to that EVC happens, which can lead to hours of outage.

The function above is used to decide if we should run sdntrace or not in an EVC, but what if we know that we should not run sdntrace, but we also know that the EVC should be redeployed (i.e, enabled, inactive and current_path is empty for inter switch EVC)

**Comment by viniarck** [comment]
> but what if we know that we should not run sdntrace, but we also know that the EVC should be redeployed

Yes, makes a lot of sense, I believe this is something to evolve as a new requirement. Currently, the redeployment in terms of consistency check only happens after it fails, but then as you pointed out it'll only help out it these circuits were being checked in the first place or not. There are multiple ways to evolve this and support it, it's multiple fold:

**1 - a)** Inactive inter EVCs without a `current_path` could also be considered candidates for a redeployment, special case. Or **b)** if we end up keeping the `current_path` the case being discussed on issue #394, then it would already be covered by sdntrace_cp since it would have a path, but then it might still be usning a non optimal path (see point 3 below)

**2 -** @Ktmi, sometime ago when fixing other issues with consistency check has also brought up a question whether it should also run for active EVCs to also provide an extra guarantee.
- This makes sense too, but we need to be careful to only run when still no recent convergence or FlowMods are going on otherwise it could result in false positives. That said, if it also runs for active circuits it's, yet another opportunity to potentially fix something if the trace fails, although, most of the cases that I can see here, `flow_manager`'s consistency check would fix it (alien flows or missing flows) indirectly, unless there was a major wrong actions mistake which is unlikely considering how much these NApps have been tested and usually flows actions are deterministic. There were also other cases of switch control-plane reachability that was discussed but last we spoke we dropped this idea and continued to handle links events only. 

**3 -** We could also assess general periodic path finding optimizations trying to find yet another best path, which most likely even for hundreds of EVCs with a decent frequency it wouldn't overwhelm `pathfinder`.  

IMO, **option 1a)** would probably be the easiest one to incrementally add support to, and then in the meantime keep evolving the discussion of what else is desirable to cover with consistency check and path optimizations/switchovers in general. In fact, Italo, this might still be worth being a new blueprint since it's heavy network related and can have major impact. Let me know what you think and if you also see other alternatives.

**Comment by italovalcy** [comment]
Great analysis, @viniarck . Yes, I believe option 1a would be the easiest and feasible solution for now. But you are right we need to further define the requirements here. Maybe a light weight blueprint would apply.
